### DISCUSSION TITLE
Stream clustering for text-based data

### DISCUSSION THREAD

**Discussion by Dennis1989** [online-ml/river#1006] [Ideas]
Hey, 

I was wondering if you are interested in supporting stream clustering algorithms working on textual inputs. There are some recent developments in the field, and we would be willing to work on a wrapper for our textClust algorithm.

Best regards,

Dennis

**Comment by MaxHalford** [comment]
Hey Dennis 👋

In theory, nothing is stopping you from building a TFIDF + clustering_model pipeline in River. But maybe there's some more specific things to put in place. Anyway, I'm not opposed to a contribution in this domain. I would however prefer if that contribution would make use of existing River building blocks, instead of being a wrapper on top of an external library. I could reconsider if the textClust is doing something River does not provide.

I hope that's clear enough! Let me know if not.

**Reply by Dennis1989** [reply]
Thank you for your fast reply Max!

Indeed there are more specific aspects when we deal with textual input and it is not just feasible to transform the input into TF-IDF (what is even the IDF if we do not know the whole corpus in a stream setting ? ;)) and use an existing stream k-means on top of that. Maybe wording was a little bit misleading. I was not really intending to write a wrapper around our implementation of textClust. I was more aiming towards extending your base Clusterer similar to what was done for the other stream algorithms like DenStream etc.

Best,

Dennis

**Reply by MaxHalford** [reply]
> what is even the IDF if we do not know the whole corpus in a stream setting ?

Indeed! As you can see in our code, we just increment a counter, which is not an exact science.

> Maybe wording was a little bit misleading. I was not really intending to write a wrapper around our implementation of textClust. I was more aiming towards extending your base Clusterer similar to what was done for the other stream algorithms like DenStream etc.

That sounds good 👍. Maybe you can start with a rough outline of what you plan to do. That way we might avoid you doing unnecessary work :)

Ideally, a good end-goal is produce an example notebook. The examples you have in your documentation look great.

In theory, what do you think about porting your textClust package to River (with or without all the functionality you have)? I know it's a big ask, it's not always easy to get "eaten" by a bigger package. But there are benefits, and it does ensure your work reaches a non-negligible amount of users and is part of a coherent ecosystem.

Feel free to push back on this :). Our goal is to become the main player in the online learning field, so we like to invite people to jump onboard the River train, but we're not zealots.

Summoning @hoanganhngo610 @raphaelsty: both of you might be interested in this discussion.

**Reply by hoanganhngo610** [reply]
@MaxHalford In fact, this is indeed what we have been looking for in a while, with a lot of requests about text clustering. As said, we can easily have a TFIDF + clustering pipeline, but it is still much cooler having a native text clustering algorithm in hand. And `textClustPy` seems to do that in a perfect fashion.

@Dennis1989 You have done great work on the package, and I extremely appreciate it. I know that the request of @MaxHalford is huge, but if you want to extend the base Clusterer and add your algorithm into River, I would certainly be extremely happy to help and try my best to make sure that it reaches the greatest audience possible out there.

The work flow is also quite simple and straightforward, as Max said. I have also recently delivered a speech on online clustering with River at KDD 2022, so an example notebook with a running algo on River would be a perfect addition to this.

**Reply by Dennis1989** [reply]
That sounds excellent @MaxHalford and @hoanganhngo610 !!

I think it shouldn't be too hard to port the algorithm to River, and I agree that there is a high chance that our algorithm will get a significant popularity boost. So for the sake of science (and application), this is the right thing to do :) 

I will spend some time preparing everything. Providing an example via a notebook indeed is probably super helpful for users. 

I would like to start with the core algorithm. To evaluate existing benchmark datasets, this should be sufficient. However, I think one advantage of our implementation is that we offer the capability to directly hook on a live Twitter Stream (which is probably the most prominent use-case for a textual stream cluster setting, at least from an application point-of-view). Maybe, in the long run, this could also be something that could be included.

**Reply by MaxHalford** [reply]
> I think it shouldn't be too hard to port the algorithm to River, and I agree that there is a high chance that our algorithm will get a significant popularity boost. So for the sake of science (and application), this is the right thing to do :)

Great, thanks for having a [positive mental attitude](https://getyarn.io/yarn-clip/1550c9c3-6a61-43bf-892a-bea157200fd3) :)

> I will spend some time preparing everything. Providing an example via a notebook indeed is probably super helpful for users.

Cool, feel free to ask questions if you're unsure.

> However, I think one advantage of our implementation is that we offer the capability to directly hook on a live Twitter Stream (which is probably the most prominent use-case for a textual stream cluster setting, at least from an application point-of-view). Maybe, in the long run, this could also be something that could be included.

I'll take care of this! I'll even add a hook to Twitch chat streams, if there's an API for it. I'll ping you in a pull request of two :)

**Reply by Dennis1989** [reply]
Hey,

I have two architectural questions that I currently face while porting the algorithm.

1) I use your BagOfWord implementation to get my list of ngrams, which is then used as the basis for micro-cluster creation. Here I have two options: Store the BoW object as a member in my textclust instance or go the pipeline route. However, as textclust requires BoW input as it is fundamentally built upon TF-IDF, it would always be the same pipeline, which means unnecessary programming lines for users.

2) Although we use TF-IDF which already prevents frequent words from having great influence, we additionally remove stop words directly each time a new observation arrives at the stream. Does it make sense to override the preprocessor step?

Best,

Dennis

**Reply by MaxHalford** [reply]
Hey hey.

> I use your BagOfWord implementation to get my list of ngrams, which is then used as the basis for micro-cluster creation. Here I have two options: Store the BoW object as a member in my textclust instance or go the pipeline route. However, as textclust requires BoW input as it is fundamentally built upon TF-IDF, it would always be the same pipeline, which means unnecessary programming lines for users.

I would suggest going with the pipeline route. I get it that you would always use a BoW before textClust, but you really never know how creative users want to be. For instance, what if users only have token counts and not texts? The pipeline route is more flexible that way, as it allows for more creative use. It also makes apparent what is going on. You can still make sure users use the BoW by providing a good example in the docstring. As an existing example in River, the linear regression should almost always be preceded by a standard scaler, but we make this happen with a pipeline too.

> Although we use TF-IDF which already prevents frequent words from having great influence, we additionally remove stop words directly each time a new observation arrives at the stream. Does it make sense to override the preprocessor step?

Good question. The short answer is that you can indeed use the `preprocessor` function to remove stop words with a function. But we could also make this explicit by adding a `stop_words` parameter. From this we can build the function ourselves in `VectorizerMixin`, which `BagOfWords` and `TFIDF` inherit from. If it's ok with you, I can take care of this :)
